ludwig/utils/calibration.py

```diff
@@ -0,0 +1,261 @@
+#! /usr/bin/env python
+# Copyright (c) 2022 Predibase, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+
+import logging
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+
+import torch
+import torch.nn as nn
+
+
+class ECELoss(nn.Module):
+    """Calculates the Expected Calibration Error of a model.
+
+    The input to this loss is the logits of a model, NOT the softmax scores.
+    This divides the confidence outputs into equally-sized interval bins.
+    In each bin, we compute the confidence gap:
+    bin_gap = | avg_confidence_in_bin - accuracy_in_bin |
+    We then return a weighted average of the gaps, based on the number
+    of samples in each bin
+    See: Naeini, Mahdi Pakdaman, Gregory F. Cooper, and Milos Hauskrecht.
+    "Obtaining Well Calibrated Probabilities Using Bayesian Binning." AAAI.
+    2015.
+    """
+
+    def __init__(self, n_bins: int = 15):
```

Comment:
if i understand correctly these bins are intervals of confidence. Any specific reason for choosing 15? Why not 20 so that they are at increments of .05?

Reply:
15 the number of bins used in all of the benchmarks in "On Calibration of Modern Neural Networks". 

BTW "Obtaining well calibrated probabilities using bayesian binning" uses 10 bins.

Added a reference to the docstring.